src/navigation/drive.py

```diff
@@ -1,76 +1,118 @@
 from __future__ import annotations
-from typing import Tuple
-
-
+from typing import Tuple, Optional
 import numpy as np
 import rospy
+from enum import Enum
 
 from geometry_msgs.msg import Twist
 from util.SE3 import SE3
 from util.np_utils import angle_to_rotate
 from util.ros_utils import get_rosparam
 
+default_constants = {
+    "max_driving_effort": 1.0,
+    "min_driving_effort": -1.0,
+    "turning_p": 10.0,
+    "max_turning_effort": 1.0,
+    "min_turning_effort": -1.0,
+    "driving_p": 1.0,
+}
+ODOM_CONSTANTS = get_rosparam("drive/odom", default_constants)
+MAP_CONSTANTS = get_rosparam("drive/map", default_constants)
+
+
+class Driver:
+    _last_angular_error: Optional[float] = None
+
+    class DriveMode(Enum):
+        TURN_IN_PLACE = 1
+        DRIVE_FORWARD = 2
+        STOPPED = 3
+
+    _driver_state: DriveMode = DriveMode.STOPPED
+
+    def _is_complete(self, target_pos: np.ndarray, rover_pose: SE3, completion_thresh: float) -> bool:
+        rover_pos = rover_pose.position
+        target_dir = target_pos - rover_pos
+        target_dist = np.linalg.norm(target_dir)
+        return bool(target_dist < completion_thresh)
+
+    def get_drive_command(
+        self: Driver,
+        target_pos: np.ndarray,
+        rover_pose: SE3,
+        completion_thresh: float,
+        turn_in_place_thresh: float,
+        in_odom: bool = False,
+    ) -> Tuple[Twist, bool]:
+
+        constants = ODOM_CONSTANTS if in_odom else MAP_CONSTANTS
+        MAX_DRIVING_EFFORT = constants["max_driving_effort"]
+        MIN_DRIVING_EFFORT = constants["min_driving_effort"]
+        MAX_TURNING_EFFORT = constants["max_turning_effort"]
+        MIN_TURNING_EFFORT = constants["min_turning_effort"]
+        TURNING_P = constants["turning_p"]
+        DRIVING_P = constants["driving_p"]
+
+        rover_dir = rover_pose.rotation.direction_vector()
+        rover_dir[2] = 0
+        rover_pos = rover_pose.position
+        rover_pos[2] = 0
+
+        target_pos[2] = 0
+        target_dir = target_pos - rover_pos
+
+        angle_error = angle_to_rotate(rover_dir, target_dir)
+
+        output = Twist(), False
+
+        if self._driver_state == self.DriveMode.STOPPED:
+            if self._is_complete(target_pos, rover_pose, completion_thresh):
+                output = Twist(), True
+            else:
```

Comment:
why switch to Turn in place based on failing is_complete? Is complete only checks position

Reply:
the reason I have is complete only check position is because the function in my view should only be responsible for driving to within a certain threshold distance of a position so if you're within that threshold of the position the angluar error is irrelevant since you're already there, the angular error is just an aid to actually get there if you're not already there yet (I guess this is more addressing your point in the overall comment). The reason I switch to turn in place is just because turn in place will switch to drive straight if we're within the angular tolerance so I didn't see the need to do the check twice, I could add it though so it wouldn't take two cycles of the state machine to get to the right spot

Reply:
yeah I think adding more transitions to get direct jumps to the right states makes things more readable for future readers even though it increases LOC